### ISSUE TITLE
Incorrect parser error when match statement is missing its ;

### ISSUE THREAD

**Issue by timthelion** [rust-lang/rust#40799]
I found an analog of bug #40006 (Forty thousand bugs, you've gotta be kidding me! :D You've put in a lot of work there!) and @sanxiyn asked me to file my own report, so here it is.

If you don't put a ; at the end of a match block then the compiler tells you very unrelated things: https://play.rust-lang.org/?gist=ab74d4c0f3dce8c095f79469bc69b072&version=nightly&backtrace=0

**Comment by oli-obk** [comment]
I don't see a parser error. I see a type error as I expected.

For posterity: the code is 

```rust
fn main() {
    match 3 {
        4 => 1,
        3 => {
            println!("Yep it maches.");
            2
        }
        _ => 2
    }
    println!("Bye!")
}
```

And the error is

```
error[E0308]: mismatched types
 --> <anon>:6:13
  |
6 |             2
  |             ^ expected (), found integral variable
  |
  = note: expected type `()`
             found type `{integer}`

error[E0308]: match arms have incompatible types
 --> <anon>:2:5
  |
2 |       match 3 {
  |  _____^ starting here...
3 | |         4 => 1,
4 | |         3 => {
5 | |             println!("Yep it maches.");
6 | |             2
7 | |         }
8 | |         _ => 2
9 | |     }
  | |_____^ ...ending here: expected integral variable, found ()
  |
  = note: expected type `{integer}`
             found type `()`
note: match arm with an incompatible type
 --> <anon>:4:14
  |
4 |           3 => {
  |  ______________^ starting here...
5 | |             println!("Yep it maches.");
6 | |             2
7 | |         }
  | |_________^ ...ending here

error: aborting due to 2 previous errors
```

**Comment by timthelion** [comment]
I dissagree.

The pragmatic reason I dissagree, is that the error is useless. It fails to identify the location and the kind of mistake that the programmer has made.

The pendantic reason I dissagree, is that the error is wrong.

````
error[E0308]: match arms have incompatible types
 --> <anon>:2:5
  |
2 |       match 3 {
  |  _____^ starting here...
3 | |         4 => 1,
4 | |         3 => {
5 | |             println!("Yep it maches.");
6 | |             2
7 | |         }
8 | |         _ => 2
9 | |     }
  | |_____^ ...ending here: expected integral variable, found ()
````

The match arms do NOT have incompatible types. Every arm of the match evaluates, eventually, to an `{integer}`. This is clear from the fact that adding a `;` fixes the problem. If there was an error in the match block itself, then adding a `;` wouldn't do any good.

A correct type error would be:


````
error[E0??]: match block's type does not match context.
 --> <anon>:2:5
  |
2 |       match 3 {
  |  _____^ starting here...
3 | |         4 => 1,
4 | |         3 => {
5 | |             println!("Yep it maches.");
6 | |             2
7 | |         }
8 | |         _ => 2
9 | |     }
  | |_____^ ...ending here: match block has type {integer}, expected ()
````

**Comment by oli-obk** [comment]
The initial error message is correct:

```
error[E0308]: mismatched types
 --> <anon>:6:13
  |
6 |             2
  |             ^ expected (), found integral variable
  |
  = note: expected type `()`
             found type `{integer}`
```

Everything else continues from there. Probably just a case of passing up a real type (the tuple) instead of `TyErr`

**Comment by estebank** [comment]
Current output has an incorrect label, but it otherwise doesn't complain about the match arms anymore:

```
error[E0308]: mismatched types
 --> src/main.rs:6:13
  |
1 | fn main() {
  |           - expected `()` because of default return type
...
6 |             2
  |             ^ expected (), found integral variable
  |
  = note: expected type `()`
             found type `{integer}`
```
